### ISSUE TITLE
dotnet restore fails on Linux 4.6.2

### ISSUE THREAD

**Issue by andschwa** [dotnet/cli#3681]
## Steps to reproduce

CentOS 7 with `kernel-ml` installed, the mainline Linux kernel at version 4.6.2.

```sh
./dotnet-install.sh -c preview -v 1.0.0-preview2-003121
~/.dotnet/dotnet restore
```

## Expected  behavior

Packages to be restored.

## Actual behavior

```
Welcome to .NET Core!
---------------------
Learn more about .NET Core @ https://aka.ms/dotnet-docs. Use dotnet --help to see available commands or go to https://aka.ms/dotnet-cli-docs.
Telemetry
--------------
The .NET Core tools collect usage data in order to improve your experience. The data is anonymous and does not include commandline arguments. The data is collected by Microsoft and shared with the community.
You can opt out of telemetry by setting a DOTNET_CLI_TELEMETRY_OPTOUT environment variable to 1 using your favorite shell.
You can read more about .NET Core tools telemetry @ https://aka.ms/dotnet-cli-telemetry.
Configuring...
-------------------
A command is running to initially populate your local package cache, to improve restore speed and enable offline access. This command will take up to a minute to complete and will only happen once.
Decompressing 100% 2175 ms
Expanding 3%Segmentation fault (core dumped)
```

## Environment data
`dotnet --info` output:

```
dotnet --info
.NET Command Line Tools (1.0.0-preview2-003121)

Product Information:
 Version:            1.0.0-preview2-003121
 Commit SHA-1 hash:  1e9d529bc5

Runtime Environment:
 OS Name:     centos
 OS Version:  7
 OS Platform: Linux
 RID:         centos.7-x64
```

This is a fresh CentOS 7 machine I made to replace the Debian Testing machine I couldn't build on in #3655.

**Comment by andschwa** [comment]
This behavior reproduced even with versions of `dotnet` (specifically 1.0.0-preview2-003119) I knew to be good on other CentOS 7 machines (where I was building fine just last week).

This does not appear to be the same as #3676.

**Comment by andschwa** [comment]
The backtrace in GDB was not useful to me:

```
#0  0x00007ffff79c66d5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffff64911b2 in CorUnix::CPalSynchronizationManager::ThreadNativeWait(CorUnix::_ThreadNativeWaitData*, unsigned int, CorUnix::ThreadWakeupReason*, unsigned int*) ()
   from /home/andrew/.dotnet/shared/Microsoft.NETCore.App/1.0.0/libcoreclr.so
#2  0x00007ffff6490df4 in CorUnix::CPalSynchronizationManager::BlockThread(CorUnix::CPalThread*, unsigned int, bool, bool, CorUnix::ThreadWakeupReason*, unsigned int*) ()
   from /home/andrew/.dotnet/shared/Microsoft.NETCore.App/1.0.0/libcoreclr.so
#3  0x00007ffff6495b81 in CorUnix::InternalWaitForMultipleObjectsEx(CorUnix::CPalThread*, unsigned int, void* const*, int, unsigned int, int) () from /home/andrew/.dotnet/shared/Microsoft.NETCore.App/1.0.0/libcoreclr.so
#4  0x00007ffff6495da2 in WaitForSingleObjectEx ()
   from /home/andrew/.dotnet/shared/Microsoft.NETCore.App/1.0.0/libcoreclr.so
#5  0x00007ffff61eded2 in CLREventBase::WaitEx(unsigned int, WaitMode, PendingSync*) ()
   from /home/andrew/.dotnet/shared/Microsoft.NETCore.App/1.0.0/libcoreclr.so
```

**Comment by eerhardt** [comment]
/cc @piotrpMSFT @livarcocc @ericstj 

It sounds like you probably don't have a required prerequisite on the machine, but I would have no idea which one...  I'm not even sure of the list of required prereqs on CentOS.

**Comment by andschwa** [comment]
As part of the prereqs I've previously identified for CentOS, I have installed:

```
curl make gcc-c++ cmake glibc libgcc libstdc++ libcurl krb5-libs libicu lldb openssl-libs libunwind libuuid zlib clang
```

This was after taking the package dependencies from the old Ubuntu `dotnet` packages (host, sdk, and tool), and finding the corresponding CentOS packages.

**Comment by eerhardt** [comment]
Can you check all the config that happens in https://github.com/dotnet/cli/blob/rel/1.0.0/scripts/docker/centos/Dockerfile?  That sets up our centos docker images for our CI.

**Comment by eerhardt** [comment]
Obviously you shouldn't need 'make' or any C++ compiler, those steps are for building the CLI and Shared Framework code itself.

**Comment by andschwa** [comment]
Checked the prereqs, already have them all. Really don't think I need the clang, lldb, and llvm packages.

```sh
$ sudo yum install unzip libunwind gettext libcurl-devel openssl-devel zlib libicu-devel
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
 * base: mirrors.syringanetworks.net
 * elrepo: ftp.osuosl.org
 * epel: mirrors.cat.pdx.edu
 * extras: mirrors.xmission.com
 * updates: mirrors.cat.pdx.edu
Package unzip-6.0-15.el7.x86_64 already installed and latest version
Package 2:libunwind-1.1-5.el7_2.2.x86_64 already installed and latest version
Package gettext-0.18.2.1-4.el7.x86_64 already installed and latest version
Package libcurl-devel-7.29.0-25.el7.centos.x86_64 already installed and latest version
Package 1:openssl-devel-1.0.1e-51.el7_2.5.x86_64 already installed and latest version
Package zlib-1.2.7-15.el7.x86_64 already installed and latest version
Package libicu-devel-50.1.2-15.el7.x86_64 already installed and latest version
Nothing to do
```

**Comment by andschwa** [comment]
@piotrpMSFT said they'd recently identified and fixed some compression library problems in CoreFX.

**Comment by eerhardt** [comment]
The version you are using is the build of .NET Core 1.0.0 RTM and the SDK Preview2 that is going to be released next week.  These should be the "golden bits".

Just to ensure everything is right:  where did you get `dotnet-install` from?  Above you are using `./` to invoke it.  Is it old/out of date somehow?  Can you ensure you are using this version:  https://github.com/dotnet/cli/blob/rel/1.0.0-preview2/scripts/obtain/dotnet-install.sh

**Comment by andschwa** [comment]
@eerhardt yes I downloaded it today, let me re-download it and test.

**Comment by andschwa** [comment]
Deleted `~/.dotnet`, re-downloaded install script, it downloaded the same [1.0.0-preview2-003121.tar.gz](https://dotnetcli.blob.core.windows.net/dotnet/preview/Binaries/1.0.0-preview2-003121/dotnet-dev-centos-x64.1.0.0-preview2-003121.tar.gz), still segfaults when expanding the initial cache.

(We have these steps automated for our CI, they should definitely be right, but I re-ran them by hand to be sure.)

**Comment by andschwa** [comment]
Here's the last bit of the `strace` before it crashes:

```
open("/home/andrew/.dotnet/sdk/1.0.0-preview2-003121/NuGet.RuntimeModel.dll", O_RDONLY) = 81
flock(81, LOCK_SH|LOCK_NB)              = 0
fcntl(81, F_SETFD, FD_CLOEXEC)          = 0
lseek(81, 0, SEEK_SET)                  = 0
read(81, "MZ\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0@\0\0\0\0\0\0\0"..., 64) = 64
lseek(81, 128, SEEK_SET)                = 128
read(81, "PE\0\0\35\375\3\0\301\37_W\0\0\0\0\0\0\0\0\360\0\" \v\2\v\0\0\0\0\0"..., 264) = 264
mmap(0x7f0ded06c000, 4096, PROT_READ, MAP_PRIVATE|MAP_FIXED, 81, 0) = 0x7f0ded06c000
mmap(0x7f0ded06d000, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED, 81, 0x1000) = 0x7f0ded06d000
mmap(0x7f0ded06e000, 57344, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED, 81, 0x2000) = 0x7f0ded06e000
mmap(0x7f0ded07c000, 4096, PROT_READ, MAP_PRIVATE|MAP_FIXED, 81, 0x10000) = 0x7f0ded07c000
mprotect(0x7f0ded046000, 4096, PROT_READ|PROT_WRITE) = 0
mprotect(0x7f0ded05d000, 4096, PROT_READ|PROT_WRITE) = 0
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e66f54000
mprotect(0x7f0e66f54000, 65536, PROT_READ|PROT_WRITE) = 0
mprotect(0x7f0dec0c0000, 4096, PROT_READ|PROT_WRITE) = 0
mprotect(0x7f0dec0c0000, 4096, PROT_READ|PROT_WRITE|PROT_EXEC) = 0
munmap(0x7f0e66f54000, 65536)           = 0
mprotect(0x7f0ded047000, 4096, PROT_READ|PROT_WRITE) = 0
mprotect(0x7f0ded048000, 4096, PROT_READ|PROT_WRITE) = 0
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e66f54000
mprotect(0x7f0e66f54000, 65536, PROT_READ|PROT_WRITE) = 0
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e5d4c1000
mprotect(0x7f0e5d4c1000, 65536, PROT_READ|PROT_WRITE) = 0
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e5d4b1000
mprotect(0x7f0e5d4b1000, 65536, PROT_READ|PROT_WRITE) = 0
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e5d4a1000
mprotect(0x7f0e5d4a1000, 65536, PROT_READ|PROT_WRITE) = 0
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e5d491000
mprotect(0x7f0e5d491000, 65536, PROT_READ|PROT_WRITE) = 0
munmap(0x7f0e66f54000, 65536)           = 0
munmap(0x7f0e5d4c1000, 65536)           = 0
munmap(0x7f0e5d4b1000, 65536)           = 0
munmap(0x7f0e5d4a1000, 65536)           = 0
munmap(0x7f0e5d491000, 65536)           = 0
mprotect(0x7f0ded05e000, 4096, PROT_READ|PROT_WRITE) = 0
mprotect(0x7f0ded049000, 4096, PROT_READ|PROT_WRITE) = 0
mprotect(0x7f0ded04a000, 4096, PROT_READ|PROT_WRITE) = 0
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e66f54000
mprotect(0x7f0e66f54000, 65536, PROT_READ|PROT_WRITE) = 0
munmap(0x7f0e66f54000, 65536)           = 0
mprotect(0x7f0ded04b000, 4096, PROT_READ|PROT_WRITE) = 0
mprotect(0x7f0dec0c1000, 4096, PROT_READ|PROT_WRITE) = 0
mprotect(0x7f0dec0c1000, 4096, PROT_READ|PROT_WRITE|PROT_EXEC) = 0
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e66f54000
mprotect(0x7f0e66f54000, 65536, PROT_READ|PROT_WRITE) = 0
munmap(0x7f0e66f54000, 65536)           = 0
mprotect(0x7f0ded04c000, 4096, PROT_READ|PROT_WRITE) = 0
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e66f54000
mprotect(0x7f0e66f54000, 65536, PROT_READ|PROT_WRITE) = 0
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e5d4c1000
mprotect(0x7f0e5d4c1000, 65536, PROT_READ|PROT_WRITE) = 0
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e5d4b1000
mprotect(0x7f0e5d4b1000, 65536, PROT_READ|PROT_WRITE) = 0
mprotect(0x7f0ded05f000, 4096, PROT_READ|PROT_WRITE) = 0
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e5d4a1000
mprotect(0x7f0e5d4a1000, 65536, PROT_READ|PROT_WRITE) = 0
munmap(0x7f0e66f54000, 65536)           = 0
munmap(0x7f0e5d4c1000, 65536)           = 0
munmap(0x7f0e5d4b1000, 65536)           = 0
munmap(0x7f0e5d4a1000, 65536)           = 0
mprotect(0x7f0ded07d000, 4096, PROT_READ|PROT_WRITE) = 0
mprotect(0x7f0ded07d000, 4096, PROT_READ|PROT_WRITE|PROT_EXEC) = 0
mprotect(0x7f0ded060000, 4096, PROT_READ|PROT_WRITE) = 0
mprotect(0x7f0ded04d000, 4096, PROT_READ|PROT_WRITE) = 0
flock(75, LOCK_UN)                      = 0
close(75)                               = 0
mprotect(0x7f0ded04e000, 4096, PROT_READ|PROT_WRITE) = 0
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e66f54000
mprotect(0x7f0e66f54000, 65536, PROT_READ|PROT_WRITE) = 0
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e5d4c1000
mprotect(0x7f0e5d4c1000, 65536, PROT_READ|PROT_WRITE) = 0
munmap(0x7f0e66f54000, 65536)           = 0
munmap(0x7f0e5d4c1000, 65536)           = 0
mprotect(0x7f0ded04f000, 4096, PROT_READ|PROT_WRITE) = 0
mprotect(0x7f0ded061000, 4096, PROT_READ|PROT_WRITE) = 0
mprotect(0x7f0ded050000, 4096, PROT_READ|PROT_WRITE) = 0
futex(0x7f0db4002f64, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f0db4002f60, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x7f0db4002f38, FUTEX_WAKE_PRIVATE, 1) = 1
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e66f54000
mprotect(0x7f0e66f54000, 65536, PROT_READ|PROT_WRITE) = 0
futex(0x7f0e659c9c54, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f0e659c9c50, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x7f0e659c9c28, FUTEX_WAKE_PRIVATE, 1) = 1
mprotect(0x7f0ded066000, 4096, PROT_READ|PROT_WRITE) = 0
futex(0x16c78cc, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x16c78c8, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x16c78a0, FUTEX_WAKE_PRIVATE, 1) = 1
munmap(0x7f0e66f54000, 65536)           = 0
futex(0x16b242c, FUTEX_WAIT_PRIVATE, 1, NULL) = -1 EAGAIN (Resource temporarily unavailable)
futex(0x16b2400, FUTEX_WAKE_PRIVATE, 1) = 0
futex(0x1826cf4, FUTEX_WAIT_PRIVATE, 1, NULL) = -1 EAGAIN (Resource temporarily unavailable)
futex(0x1826cc8, FUTEX_WAKE_PRIVATE, 1) = 0
futex(0x1826bc4, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x1826bc0, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x1826b98, FUTEX_WAKE_PRIVATE, 1) = 1
mprotect(0x7f0ded0da000, 4096, PROT_READ|PROT_WRITE) = 0
futex(0x7f0e659c9c54, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f0e659c9c50, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x7f0e659c9c28, FUTEX_WAKE_PRIVATE, 1) = 1
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e5d4c1000
mprotect(0x7f0e5d4c1000, 65536, PROT_READ|PROT_WRITE) = 0
futex(0x7f0e659c9c54, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f0e659c9c50, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x7f0e659c9c28, FUTEX_WAKE_PRIVATE, 1) = 1
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e66f54000
mprotect(0x7f0e66f54000, 65536, PROT_READ|PROT_WRITE) = 0
futex(0x7f0e659c9c54, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f0e659c9c50, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x7f0e659c9c28, FUTEX_WAKE_PRIVATE, 1) = 1
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e5d4b1000
mprotect(0x7f0e5d4b1000, 65536, PROT_READ|PROT_WRITE) = 0
futex(0x7f0e659c9c54, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f0e659c9c50, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x7f0e659c9c28, FUTEX_WAKE_PRIVATE, 1) = 1
munmap(0x7f0e5d4c1000, 65536)           = 0
futex(0x7f0e659c9c54, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f0e659c9c50, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x7f0e659c9c28, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x7f0e659c9c54, FUTEX_WAIT_PRIVATE, 13, NULL) = -1 EAGAIN (Resource temporarily unavailable)
futex(0x7f0e659c9c28, FUTEX_WAKE_PRIVATE, 1) = 0
munmap(0x7f0e66f54000, 65536)           = 0
munmap(0x7f0e5d4b1000, 65536)           = 0
futex(0x7f0e659c9c54, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f0e659c9c50, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x7f0e659c9c28, FUTEX_WAKE_PRIVATE, 1) = 1
mprotect(0x7f0deb56b000, 4096, PROT_READ|PROT_WRITE) = 0
mprotect(0x7f0deb56b000, 4096, PROT_READ|PROT_WRITE|PROT_EXEC) = 0
futex(0x7f0e659c9c54, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f0e659c9c50, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x7f0e659c9c28, FUTEX_WAKE_PRIVATE, 1) = 1
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e5d4c1000
mprotect(0x7f0e5d4c1000, 65536, PROT_READ|PROT_WRITE) = 0
futex(0x7f0e659c9c54, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f0e659c9c50, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x7f0e659c9c28, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x7f0e659c9c54, FUTEX_WAIT_PRIVATE, 21, NULL) = -1 EAGAIN (Resource temporarily unavailable)
futex(0x7f0e659c9c28, FUTEX_WAKE_PRIVATE, 1) = 0
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e66f54000
mprotect(0x7f0e66f54000, 65536, PROT_READ|PROT_WRITE) = 0
futex(0x7f0e659c9c54, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f0e659c9c50, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x7f0e659c9c28, FUTEX_WAKE_PRIVATE, 1) = 1
munmap(0x7f0e5d4c1000, 65536)           = 0
futex(0x7f0e659c9c54, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f0e659c9c50, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x7f0e659c9c28, FUTEX_WAKE_PRIVATE, 1) = 1
munmap(0x7f0e66f54000, 65536)           = 0
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e5d4b1000
mprotect(0x7f0e5d4b1000, 65536, PROT_READ|PROT_WRITE) = 0
futex(0x7f0e659c9c54, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f0e659c9c50, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x7f0e659c9c28, FUTEX_WAKE_PRIVATE, 1) = 1
mmap(NULL, 65536, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0e5d4c1000
mprotect(0x7f0e5d4c1000, 65536, PROT_READ|PROT_WRITE) = 0
futex(0x7f0e659c9c54, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f0e659c9c50, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x7f0e659c9c28, FUTEX_WAKE_PRIVATE, 1) = 1
munmap(0x7f0e5d4b1000, 65536)           = 0
futex(0x7f0e659c9c54, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f0e659c9c50, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x7f0e659c9c28, FUTEX_WAKE_PRIVATE, 1) = 1
munmap(0x7f0e5d4c1000, 65536)           = 0
sched_yield()                           = 0
futex(0x169eab4, FUTEX_WAIT_PRIVATE, 113, NULL) = ? <unavailable>
+++ killed by SIGSEGV (core dumped) +++
Segmentation fault (core dumped)
```

**Comment by andschwa** [comment]
I found it:

This segmentation fault occurs when running with the mainline kernel, 4.6.2, but not with 3.10.0 (CentOS 7's default).

Now *why* decompression fails on an up-to-date kernel is beyond me, but I'll change this name of this issue.

**Comment by eerhardt** [comment]
Awesome!  Thanks for tracking it down, @andschwa. Adding some CoreCLR and CoreFX folks who may have a better idea of the issue on an updated CentOS kernel.

/cc @gkhanna79 @ellismg @sergiy-k @jkotas @janvorli

**Comment by andschwa** [comment]
Hey wouldn't you know it: I can reproduce this on Ubuntu 14.04 with the 4.6.2 image from [kernel.ubuntu.com](http://kernel.ubuntu.com/~kernel-ppa/mainline/v4.6.2-yakkety/). At least it's consistent!

This also tells me that it doesn't repro on 4.2.0 either.

**Comment by jasonwilliams200OK** [comment]
@andschwa, do you have similar symptoms with updated kernel as http://unix.stackexchange.com/q/253903? Then these might be related:

Answer: http://unix.stackexchange.com/a/255603
Announcement: https://github.com/systemd/systemd/blob/master/NEWS#L60
PR: https://github.com/systemd/systemd/pull/1239

**Comment by andschwa** [comment]
I'm not seeing the same errors as the Stack Exchange question. I have this in dmesg when it crashes:

```
[   85.928175] traps: dotnet[3374] general protection ip:7f51d0f5f815 sp:7f51a9ff81e0 error:0
```

I've ruled out the Hypervisor and any extensions, as the CentOS VM is on VirtualBox (with guest additions) and the Ubuntu VM is on Hyper-V (with kernel LIS drivers).

Running through valgrind didn't give me much, `dotnet restore` seemed to hang. Seeing what I can do about.

Attempting to get a coredump.

**Comment by andschwa** [comment]
Coredumps obtained for CentOS (and Debian), emailing them to you @ellismg.

Fortunately, as I know how to reproduce the problem, I can downgrade my kernel and move on with work. Nonetheless, we should figure out why 4.6.x kernels and `dotnet` don't play well. Which issue should we keep (pretty sure they're the same problem now)?

**Comment by ellismg** [comment]
I vote we keep this one.

**Comment by janvorli** [comment]
Let me take a look.

**Comment by janvorli** [comment]
I have installed the 4.6.2 kernel to my Ubuntu VM and found it has nothing to do with the cli. The runtime itself doesn't work on that linux kernel. Even attempt to build coreclr repo fails with SIGSEGV in GC when building mscorlib.dll / System.Private.CoreLib.dll.
I have created issue in coreclr assigned to myself to track it:
https://github.com/dotnet/coreclr/issues/6016

**Comment by klinkby** [comment]
I can confirm that Debian stretch (4.6.0-1-amd64) is also affected with 1.0.0-preview2-003121.

**Comment by jasonwilliams200OK** [comment]
@klinkby, PR that fixes the issue: https://github.com/dotnet/coreclr/pull/6027.

**Comment by abrodersen** [comment]
Any chance the fix can be pushed out in a hotfix patch? This is currently preventing us from upgrading to 1.0

**Comment by cruz82** [comment]
Having the same issue on Fedora 24 after fixing the libicu version issues.
Is there a workaround for this?

> $ dotnet --version
> 1.0.0-preview3-003208
> $ dotnet new
>
> Decompressing 100% 1951 ms
> Expanding 4%/usr/local/bin/dotnet: line 5: 11895 Segmentation fault      (core dumped) /opt/dotnet/dotnet 
> $ dotnet --info
> .NET Command Line Tools (1.0.0-preview3-003208)
>
> Product Information:
> Version:            1.0.0-preview3-003208
> Commit SHA-1 hash:  76058bea60
>
> Runtime Environment:
> OS Name:     fedora
> OS Version:  24
> OS Platform: Linux
> RID:         fedora.24-x64

**Comment by ellismg** [comment]
@cruz82 What version of the kernel are you using?

**Comment by cruz82** [comment]
I'm on 4.6.3-300

**Comment by ellismg** [comment]
There was a bug in 1.0.0, that manifests itself as a crash like the above on recent kernels (IIRC 4.6.0+).  We have a fix in the runtime but no official release of it yet.  If you want to rebuild the CoreCLR part from source I can show you the commit to cherry-pick.

**Comment by tanordheim** [comment]
We're struggling with this also - we're seeing some instabilities in some apps running in Docker here after the underlying hosts (CoreOS) received some updates to the kernel recently. We're unable to run new builds right now (also happening in the containers) and already built apps are crashing fairly often. We can rebuild the CLR from source based on the latest preview2 if we can get some more information about which commits fixes this.

**Comment by cruz82** [comment]
Sure, could try building with the appropriate patch or better yet, is there a runnable branch with the fix in place?

**Comment by shahid-pk** [comment]
i think this was the commit in coreclr https://github.com/dotnet/coreclr/pull/6027  @ellismg can confirm.

**Comment by janvorli** [comment]
@shahid-pk yes, that was the fix.

**Comment by tanordheim** [comment]
For someone who's not really familiar with building this platform from source - any steps in some documentation on how I can go about fixing this? I have a built of coreclr here with this patched onto the 1.0.0-branch but I can't really figure out how to go about actually making that into a fix we can deploy into our currently deployed .NET core apps that are experiencing some failures at the moment. Do we need to rebuild everything (coreclr+corefx+cli) and use that, or can we replace some running parts of a system with stuff from this coreclr build and potentially fix this problem until we get a new official release containing this fix?

**Comment by shahid-pk** [comment]
@tanordheim you will only need to replace binaries you get from coreclr-1.0.0-branch+(https://github.com/dotnet/coreclr/pull/6027), that is if your application is self contained and not using shared host/framework/runtime installed system wide in which case you will have to build all coreclr+corefx+cli.

Note : I am not a member of .net team. My words should be taken as an advise which should work. Team members may advise differently.

**Comment by jasonwilliams200OK** [comment]
@tanordheim, dotnet/coreclr#6027 is a link of PR, the in-tree commit of master branch is dotnet/coreclr@56ab756. 

You most probably only need to replace mscorlib, libcoreclr, and system.private.corlib (in total 5 items, see below) from the build output and replace with the installed one. Here is how:

```sh
# this is std Bourne shell
git clone https://github.com/dotnet/coreclr -b release/1.0.0

cd coreclr

git cherry-pick 56ab756

build.sh release

# it may ask you to install cmake, libunwind etc. if you haven't already,

# On successful build, time to copy stuff:

# 1)

# make backup of existing libcoreclr.so in the installed runtime dir first:
mv $(dirname $(which dotnet))/shared/Microsoft.NETCore.App/1.0.0/libcoreclr.so \
$(dirname $(which dotnet))/shared/Microsoft.NETCore.App/1.0.0/libcoreclr.so.old
# ^ this way, you can always move `libcoreclr.so.old` back to `libcoreclr.so`

# copy newly built libcoreclr.so over to the installed runtime dir
cp bin/Product/Linux.x64.Release/libcoreclr.so \
$(dirname $(which dotnet))/shared/Microsoft.NETCore.App/1.0.0/

# 2)

# similarly, make backup of existing mscorlib.dll in the installed runtime dir:
mv $(dirname $(which dotnet))/shared/Microsoft.NETCore.App/1.0.0/mscorlib.dll \
$(dirname $(which dotnet))/shared/Microsoft.NETCore.App/1.0.0/mscorlib.dll.old

# copy newly built mscorlib.dll over to the installed runtime dir
cp bin/Product/Linux.x64.Release/mscorlib.dll \
$(dirname $(which dotnet))/shared/Microsoft.NETCore.App/1.0.0/

# repeat for the following:

# 3) bin/Product/Linux.x64.Release/mscorlib.ni.dll
# 4) bin/Product/Linux.x64.Release/System.Private.CoreLib.dll
# 5) bin/Product/Linux.x64.Release/System.Private.CoreLib.ni.dll
```

CLI team at some point announced the nightly builds, but apparently that plan is not on board any more..

**Comment by cruz82** [comment]
@jasonwilliams200OK @ellismg 
Tried building the release/1.0.0 branch but getting a

>/coreclr/src/pal/src/cruntime/mbstring.cpp:182:20: error: use of undeclared identifier 'min'
>            ret += min(count, strlen((const char*)string));
>                   ^
>1 error generated.
>src/pal/src/CMakeFiles/coreclrpal.dir/build.make:182: recipe for target >'src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/mbstring.cpp.o' failed
>make[2]: *** [src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/mbstring.cpp.o] Error 1
>make[2]: *** Waiting for unfinished jobs....

This is even before I use the cherry-pick.
I've tried building on master and that goes to the end but fails to generate the following files as listed in your comment:
* mscorlib.dll
* mscorlib.ni.dll
* System.Private.CoreLib.dll
* System.Private.CoreLib.ni.dll

**Comment by jasonwilliams200OK** [comment]
@cruz82, you probably need to apply this patch https://github.com/dotnet/coreclr/pull/5304/files.

`git cherry-pick 41912e3`

(updated the comment above with same)

Related thread: https://github.com/dotnet/coreclr/issues/5006

**Comment by cruz82** [comment]
@jasonwilliams200OK I've got it building with the cherry pick 41912e3

That produces the libcoreclr.so but now I seem to be missing some items to build the rest of it.

>Running: "/coreclr/Tools/corerun" "/coreclr/Tools/MSBuild.exe" "/coreclr/build.proj" /v:minimal >/t:GenerateVersionSourceFile >/p:NativeVersionSourceFile=/coreclr/bin/obj/Linux.x64.Release/version.cpp >/p:GenerateVersionSourceFile=true /v:minimal 
>./build.sh: line 158: /coreclr/Tools/corerun: No such file or directory

* Tools
    * 1.0.25-prerelease-00510-02
        * project.json
    * dotnetcli
        * dotnet.tar (seems incomplete as it's 215 bytes and not a valid tar)

Do I need the appropriate CLI package in that folder for it to build the managed components portion?

**Comment by cruz82** [comment]
Apparently I only need the _libcoreclr.so_ and the other files can be ignored.

Thanks @jasonwilliams200OK @ellismg

**Comment by joshfree** [comment]
related coreclr issue tracked here: https://github.com/dotnet/coreclr/issues/6604

**Comment by devlead** [comment]
As this is fixed in dotnet/coreclr#6604 & released with v1.0.4 will we see a preview4 soon of the CLI? Several CI servers have now updated to 4.6.2+ i.e. GitLabs now fail on all CoreCLR builds.

**Comment by jasonwilliams200OK** [comment]
> will we see a preview4 soon of the CLI?

wait, is preview3 already released? I thought preview 2 is the latest one: https://github.com/dotnet/cli/releases/tag/v1.0.0-preview2 :open_mouth:

**Comment by devlead** [comment]
preview3 is what you get if you run latest version with, don't know if it's "released" though :wink:
https://github.com/dotnet/cli/blob/rel/1.0.0/scripts/obtain/dotnet-install.sh

**Comment by gkhanna79** [comment]
> As this is fixed in dotnet/coreclr#6604 & released with v1.0.4 will we see a preview4 soon of the CLI?

I am not sure of an upcoming CLI preview release but we are trying to determine if we can have a patch update of .NET Core Redist that will carry this change (as it should not require any app update).

@leecow is working on that plan.

**Comment by jasonwilliams200OK** [comment]
Thanks @gkhanna79! IMO, this issue deserves <kbd>bug</kbd> and/or <kbd>require-hotfix</kbd> label. :)

@devlead, I used that script in our CI script, but now (and as of our moving-forward plan) we are getting CLI from the released URLs https://github.com/nunit/nunit.portable.agent/pull/12 -> https://www.microsoft.com/net/download#core (short: http://dot.net/core) :sunglasses:

**Comment by NinoFloris** [comment]
@gkhanna79 @leecow Good news about the fix but, how is that plan coming along?

**Comment by shanselman** [comment]
I'm seeing this on Linode. Very frustrating. What's the plan here, this is two months old.

```
root@ubuntu:~# uname -r
4.6.3-x86_64-linode70
```

**Comment by DHowett** [comment]
Incidentally, the suggested workaround for solving the issue requires a
separate machine to build the replacement library because the dotnet build
system is self-hosting.
That is, the build for the update fails because it attempts to `dotnet
restore`. :P

**Comment by devlead** [comment]
Would be nice if there at least was an nightly tar.gz I could pickup, currently a bunch of CI services and container services is a total no-go with CoreCLR, and with the ones that currently work it feels like a ticking time bomb until they patch and things stop working there too. As containers share host kernel it ain't much you can do in that scenario except change vendor, and that's rarely an quickfix/viable option.

**Comment by robertmiles3** [comment]
I'm with @shanselman...a fresh Linode of Ubuntu 16.04 and can't run my core web app because of this. Is there a timeframe to get this fix out in the wild without having to do source building and commit cherry-picking?

**Comment by jasonwilliams200OK** [comment]
+:100: for the next preview build, I am ALL in. But would be AWESOME if this #4074 goes in as well, so maybe wait a little longer? :blush:

**Comment by Yantrio** [comment]
Any update on when this will be fixed?

**Comment by Bhaal22** [comment]
even with @jasonwilliams200OK  trick I have the issue running *dotnet restore*
The fist commitID provided by @jasonwilliams200OK  is already included in the release/1.0.0

john@roth$ git cherry-pick 56ab756
error: could not apply 56ab756... Fix Windows context to Unix context translation on AMD64 (#6027)

Here is the conflict :
```
              g_activationFunction(&winContext);
++<<<<<<< HEAD
 +
++=======
++>>>>>>> 56ab756... Fix Windows context to Unix context translation on AMD64 (#6027)
```

Still need to apply the second commitId to fix compilation issue with min symbol not found.

But then ... restore still seg fault.

**Comment by narmontas** [comment]
@Bhaal22 the conflict is easy to fix. I replaced only _libcoreclr.so_ as per @cruz82 suggestion and segfault disappeared.

**Comment by Bhaal22** [comment]
@narmontas yeah the conflict is a noop at this stage. This why I said the commit is already in the branch unless this empty line ... but we dont care.

**Comment by jasonwilliams200OK** [comment]
@Bhaal22, yeah https://github.com/dotnet/coreclr/commit/56ab756 in master was landed in release/1.0.0 on Jul 29 via https://github.com/dotnet/coreclr/commit/e4d5001206fde605a255669079c50b89e5f32324. I wrote that comment on Jul 23. :wink: 
Also there is other stuff been cherry-picked (or merged by someone else with new commit message.. which to my knowledge is not a great idea.. esp. overwriting the authorship part and then causing the disparity across the branch trees) since then: https://github.com/dotnet/coreclr/commits/release/1.0.0.

**Comment by Bhaal22** [comment]
@jasonwilliams200OK not a problem i do not blame you :) really thank you for your inputs. on my system, even after the "cherry-pick 41912e3" (this one fix the *min* compilation issue). 
on my system kernel 4.6.0-1 I still have segmentation fault on restore.

of course I copied libcoreclr

```
cp libcoreclr.so ~/dev/tools/dotnet/shared/Microsoft.NETCore.App/1.0.0/libcoreclr.so
```

and 

```
john@roth$ dotnet --version
1.0.0-preview3-003223
```

**Comment by shahid-pk** [comment]
@Bhaal22 1.0.0-preview3-003223 is not rel 1.0.0 it has many other new bits in coreclr and corefx , you are running an unstable build which most probably have other issues. This version 1.0.0-preview2-003121 has coreclr and corefx version 1.0.0 , you should be applying the patch to this version.

**Comment by Bhaal22** [comment]
well .. still have the issue even with preview2 ... hmmm

**Comment by NinoFloris** [comment]
And this here is exactly why we have been pleading for an official fix. It cannot be reasonably expected for normal users to compile some native app from source and repackage it with all the tooling to create a fix for themselves.

Every day this stays unresolved it gets more embarrassing. And it starts to become writing on the wall for our trust concerning swift resolve the next time an issue like this comes up. 

It has been more than a few weeks ago since I got the coreclr guys to push the fix into the main branch and it has been more than 3 months ago since similar issues started to crop up.

Is this what we'll have to get used to when it comes to support for a now opensource .net? As we as a community cannot change the links on http://dot.net to point to a new release, or make one for that matter. Either do full public community governance or be prepared to do this kind of work yourself within a normal timeframe.

It amazes me, really...

**Comment by gkhanna79** [comment]
There are official nuget packages for CoreCLR that carry this fix - @leecow Can you please share the steps on how to consume them?

**Comment by jasonwilliams200OK** [comment]
> As we as a community cannot change the links on http://dot.net to point to a new release

Also would have been nice to have Oracle Linux listed at https://www.microsoft.com/net/download#core by now, which is a supported platform and shares the same package as CentOS (https://github.com/aspnet/EntityFramework/issues/4293#issuecomment-229142303). Can we have a repo where we can send a PR for that website, when the team misses out something? :)

#LetUsPort.NETCoreTo[BananaUnix](https://github.com/dotnet/cli/issues/1532)And[GoCrazyWithIt](https://github.com/dotnet/coreclr/issues/917)_AKA_Freedom! :rocket: :guitar: :tada:

**Comment by gkhanna79** [comment]
> nice to have Oracle Linux listed at https://www.microsoft.com/net/download#core by now

@ellismg Thoughts on updating this?

**Comment by Petermarcu** [comment]
We're actively working on unblocking and much of the sentiment here is well founded. Very sorry for all the delays on this. We are definitely interested in making sure the community is able to get new builds with fixes and unblock and we're invested in making that a reality. Talk to @ellismg for more details.

For this issue, we're going to very shortly publish steps for how to manually patch your install with the binaries from the released nuget packages. That is the first step to unblocking folks. We're actively building those instructions now.

After that, we're going to get an updated installer people can use that will install these updated bits and the CLI will start running on those.

Last step will be updating the installers at dot.net to have the fix by default.

**Comment by yetanotherchris** [comment]
This is happening on Gitlab CI Runners (using their Docker shared runners):

Running with gitlab-ci-multi-runner 1.4.2 (bcc1794)
			Using Docker executor with image microsoft/dotnet:latest ...
			Pulling docker image microsoft/dotnet:latest ...
			Running on runner-8a2f473d-project-1521505-concurrent-0 via runner-8a2f473d-machine-1472240490-034c3d16-digital-ocean-4gb...
			Cloning repository...
			Cloning into '/builds/yetanotherchris/Tanglenet'...
			Checking out 81578ea3 as master...
			$ dotnet restore
			log  : Restoring packages for /builds/yetanotherchris/Tanglenet/src/Tanglenet.Web/project.json...
			log  : Restoring packages for /builds/yetanotherchris/Tanglenet/src/Tanglenet.Tests/project.json...
			/bin/bash: line 42:     9 Segmentation fault      (core dumped) dotnet restore
			ERROR: Build failed: exit code 1

**Comment by yetanotherchris** [comment]
The `.gitlab-ci.yml` file:

job_build_dotnet:
	       stage: build
	       image: microsoft/dotnet:latest
	       script:
	         - dotnet restore

**Comment by Petermarcu** [comment]
Here is a [link](https://github.com/dotnet/core/blob/master/release-notes/1.0/manual-shared-update.md) to the first pass on instructions to manually update .NET Core using the NuGet packages.

These instructions will get simplified in the next couple days to just point to zips that can be extracted to the right location. We're working on those now.

Shortly after that, we'll update the link to the installers to updated installers.

**Comment by Petermarcu** [comment]
@yetanotherchris , @MichaelSimons yes, we'll need to update the image at microsoft/dotnet:latest as well. @MichaelSimons can give an update on when that can happen.

**Comment by viktorbk** [comment]
**For ubuntu 14.04:**
mkdir tmp-update && cd tmp-update
wget https://www.nuget.org/api/v2/package/runtime.ubuntu.14.04-x64.Microsoft.NETCore.Runtime.CoreCLR/1.0.4
mv 1.0.4 tmp.zip
unzip tmp.zip
sudo cp ./runtimes/ubuntu.14.04-x64/lib/netstandard1.0/* /usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.0
sudo cp ./runtimes/ubuntu.14.04-x64/native/* /usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.0
**then dotnet restore works**

**Comment by jasonwilliams200OK** [comment]
@viktorbk, nice! Maybe use `which` to make script install-dir agnostic, like:

```sh
d1=$(dirname $(which dotnet))/shared/Microsoft.NETCore.App/1.0.0

sudo cp ./runtimes/ubuntu.14.04-x64/lib/netstandard1.0/* ./runtimes/ubuntu.14.04-x64/native/* $d1
# cp is capable of handling multi-source, single-target
````

**Comment by yetanotherchris** [comment]
Thanks @viktorbk that has helped resolve the issue/work around it. If anyone uses Gitlab (or maybe other CI services that use YAML), this is my definition:

    job_build_dotnet:
           stage: build
           image: microsoft/dotnet:latest
           script:
             - apt-get update
             - apt-get install unzip
             - mkdir tmp-update && cd tmp-update
             - wget https://www.nuget.org/api/v2/package/runtime.ubuntu.14.04-x64.Microsoft.NETCore.Runtime.CoreCLR/1.0.4
             - mv 1.0.4 tmp.zip
             - unzip tmp.zip
             - cp ./runtimes/ubuntu.14.04-x64/lib/netstandard1.0/* /usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.0
             - cp ./runtimes/ubuntu.14.04-x64/native/* /usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.0
             - cd ..
             - dotnet restore
             - dotnet publish src/FakeNameHere.Web
           artifacts:
             expire_in: 12 hrs
             paths:
              - src/FakeNameHere.Web/bin/Debug/netcoreapp1.0/publish/
            
    job_build_docker:
           stage: deploy
           image: docker:latest
           services:
            - docker:dind
           script:
            - docker login -u gitlab-ci-token -p $CI_BUILD_TOKEN registry.gitlab.com/yetanotherchris/FakeNameHere 
            - docker build -t registry.gitlab.com/yetanotherchris/FakeNameHere:latest .
            - docker push registry.gitlab.com/yetanotherchris/FakeNameHere:latest
           dependencies:
            - job_build_dotnet

**Comment by Petermarcu** [comment]
Awesome. Thanks for sharing scripts and verifying things are working!

**Comment by leecow** [comment]
Uploaded a script that makes the manual process reasonably automatic. Complete installer updates are expected to be out mid-September. 

https://github.com/dotnet/core/blob/master/release-notes/1.0/install-1.0.1.sh

**Comment by naamunds** [comment]
I've verified that this is fixed on Ubuntu 14.04 with kernel 4.6.2 in the .NET Core 1.0.1 update released today (SDK version 1.0.0-preview2-003131).  It can be installed from https://www.microsoft.com/net/core.

**Comment by Bhaal22** [comment]
Same issue on debian.

```
john@roth$ uname -a
Linux roth 4.6.0-1-amd64 #1 SMP Debian 4.6.4-1 (2016-07-18) x86_64 GNU/Linux
```

```
john@roth$ dotnet --info
.NET Command Line Tools (1.0.0-preview2-003131)

Product Information:
 Version:            1.0.0-preview2-003131
 Commit SHA-1 hash:  635cf40e58

Runtime Environment:
 OS Name:     debian
 OS Version:  
 OS Platform: Linux
 RID:         debian.-x64
```

core dump backtrace:

```
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `dotnet restore'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f177e2bdd4d in ?? () from /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0
[Current thread is 1 (Thread 0x7f175f134700 (LWP 4147))]
(gdb) bt
#0  0x00007f177e2bdd4d in ?? () from /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0
#1  0x00007f177e2b8c6b in X509_verify_cert () from /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0
#2  0x00007f179fee7925 in ?? ()
#3  0x00007f175f132c00 in ?? ()
#4  0xbafbcb4b7ab40f52 in ?? ()
#5  0x000000000003ac21 in ?? ()
#6  0x00007f1818856f90 in vtable for InlinedCallFrame () from /home/john/dev/tools/dotnet/shared/Microsoft.NETCore.App/1.0.1/libcoreclr.so
#7  0x00007f175f1335e8 in ?? ()
#8  0x00007f17a000cec0 in ?? ()
#9  0x00007f17a000cec0 in ?? ()
#10 0x00007f175f132c00 in ?? ()
#11 0x00007f179fee7925 in ?? ()
#12 0x00007f175f132ca0 in ?? ()
#13 0x00007f17a000cec0 in ?? ()
#14 0x00007f17803b9e60 in ?? ()
#15 0x00007f1774087ef0 in ?? ()
#16 0x0000000000000000 in ?? ()
(gdb) quit
```

**Comment by svick** [comment]
@Bhaal22 Not sure if it's the cause of your issue, but it looks like `dotnet` can't figure out which version of debian you're using.

Which version of debian *are* you using?

**Comment by Bhaal22** [comment]
@svick 

```
john@roth$ lsb_release -da
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux testing (stretch)
Release:        testing
Codename:       stretch
```

**Comment by naamunds** [comment]
@janvorli Does the segmentation fault that @Bhaal22 is still seeing on Debian Stretch look like the same issue, or should a new issue be logged for it?

**Comment by janvorli** [comment]
@naamunds The former issue manifested in a very random way - the stack segment register was getting overwritten for threads at random times (when GC was starting) and took effect after the GC finished and each thread was resumed and so the issue was always at a different place.
I would say that if the issue @Bhaal22 sees has always the same bottom of the call stack, it is a different problem.

**Comment by gkhanna79** [comment]
@Bhaal22 Can you please export COREHOST_TRACE=1, attempt the repro and share the following:

1) Log containing the trace output
2) Crash dump for the crash.

**Comment by Bhaal22** [comment]
Fore some reason (probably a system update), now its fine. 
My bad ...

Close the issue then.

**Comment by cobrafast** [comment]
I'm still running into this problem with `dotnet restore` on my Debian box with the 1.1 package as described at https://www.microsoft.com/net/core#linuxdebian.

```
Program received signal SIGSEGV, Segmentation fault.
0x00007fffefb88d6d in ?? () from /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0
(gdb) backtrace
#0  0x00007fffefb88d6d in ?? () from /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0
#1  0x00007fffefb83c8b in X509_verify_cert () from /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0
#2  0x00007fff7ce5e2c5 in ?? ()
#3  0x00007fff3e209be0 in ?? ()
#4  0xbafbcb4b7ab40f52 in ?? ()
#5  0x000000001fd117cb in ?? ()
#6  0x00007ffff67bf0f0 in vtable for InlinedCallFrame () from /opt/dotnet/shared/Microsoft.NETCore.App/1.1.1/libcoreclr.so
#7  0x00007fff3e20a5f8 in ?? ()
#8  0x00007fff7d667cc0 in ?? ()
#9  0x00007fff7d667cc0 in ?? ()
#10 0x00007fff3e209be0 in ?? ()
#11 0x00007fff7ce5e2c5 in ?? ()
#12 0x00007fff3e209c80 in ?? ()
#13 0x00007fff7d667cc0 in ?? ()
#14 0x00007fff58114100 in ?? ()
#15 0x00007fff500ddfc0 in ?? ()
#16 0x0000000000000000 in ?? ()
```

```
.NET Command Line Tools (1.0.3)

Product Information:
 Version:            1.0.3
 Commit SHA-1 hash:  37224c9917

Runtime Environment:
 OS Name:     debian
 OS Version:  8
 OS Platform: Linux
 RID:         debian.8-x64
 Base Path:   /opt/dotnet/sdk/1.0.3
```

```
Distributor ID: Debian
Description:    Debian GNU/Linux 8.7 (jessie)
Release:        8.7
Codename:       jessie
```

```
Linux 4.9.0-2-amd64 #1 SMP Debian 4.9.18-1 (2017-03-30) x86_64
```

`COREHOST_TRACE=1` resulted in a lot of text: https://pastebin.com/C7gtMr41

I'm not sure how to create or where to find crash dumps.

I have installed a few backport/sid packages as I need them for other things running on the same machine. That said, running `apt-get upgrade` didn't do anything for this particular problem.

Here are some currently installed packages that might be relevant:

```
ii  libcurl3:amd64                                              7.52.1-5                             amd64
ii  libcurl3:i386                                               7.52.1-5                             i386
ii  libcurl3-gnutls:amd64                                       7.52.1-5                             amd64
ii  libcurl3-gnutls:i386                                        7.52.1-5                             i386
ii  libcurl4-gnutls-dev:amd64                                   7.52.1-5                             amd64
ii  libssl-dev:amd64                                            1.0.1t-1+deb8u6                      amd64
ii  libssl-doc                                                  1.0.1t-1+deb8u6                      all
ii  libssl0.9.8                                                 0.9.8o-4squeeze14                    amd64
ii  libssl1.0.0:amd64                                           1.0.1t-1+deb8u6                      amd64
ii  libssl1.0.0:i386                                            1.0.1t-1+deb8u6                      i386
ii  libssl1.0.2:amd64                                           1.0.2k-1                             amd64
ii  libssl1.0.2:i386                                            1.0.2k-1                             i386
ii  libssl1.1:amd64                                             1.1.0e-1                             amd64
```

**Comment by sekjun9878** [comment]
I am also running into this problem on Debian Stretch. So much for getting started with learning C# :(

**Comment by eerhardt** [comment]
@sekjun9878 - please see this conversation for Debian Stretch: https://github.com/dotnet/core/issues/649#issuecomment-301111400

**Comment by sekjun9878** [comment]
@eerhardt Thanks for the reference, but it's not the same issue. I can run `dotnet new console -o hwapp` successfully. It's only `dotnet restore` that is segfaulting.

**Comment by eerhardt** [comment]
The general thread talks about that .NET Core 2.0 doesn't support Debian Stretch unless you install OpenSSL 1.0.x.  Do you have that installed on the machine?  Debian Stretch comes with OpenSSL 1.1, which .NET Core doesn't support yet.

So you'll get a segfault anywhere that tries using HTTPS or SSL.

**Comment by cobrafast** [comment]
@eerhardt as you can see from my post, there were various versions of libssl installed on my machine, two of which are 1.0.x versions in 64 and 32 bit variants, but it still segfaulted.

Here are the relevant files those packages result in:

```
/usr/lib/x86_64-linux-gnu/libssl3.so.1d -> libssl3.so
/usr/lib/x86_64-linux-gnu/libssl.so -> libssl.so.1.0.0
/usr/lib/x86_64-linux-gnu/libssl3.so
/usr/lib/x86_64-linux-gnu/libssl.a
/usr/lib/x86_64-linux-gnu/libssl.so.1.0.0
/usr/lib/x86_64-linux-gnu/pkgconfig/libssl.pc
/usr/lib/x86_64-linux-gnu/libssl.so.1.0.2
/usr/lib/x86_64-linux-gnu/libssl.so.1.1
/usr/lib/i386-linux-gnu/libssl3.so.1d -> libssl3.so
/usr/lib/i386-linux-gnu/i586/libssl.so.1.0.0
/usr/lib/i386-linux-gnu/libssl3.so
/usr/lib/i386-linux-gnu/libssl.so.1.0.0
/usr/lib/i386-linux-gnu/i686/cmov/libssl.so.1.0.0
/usr/lib/i386-linux-gnu/libssl.so.1.0.2
/usr/lib/libssl.so.0.9.8
```

Maybe `dotnet` is using the wrong one?

**Comment by omajid** [comment]
Can you do an `strace dotnet` or `LD_DEBUG=libs dotnet` to see what it's loading?

**Comment by cobrafast** [comment]
Sure: https://pastebin.com/s0NqFQDD

As far as I can tell it does load `libssl.so.1.0.0` but segfaults anyways.

**Comment by eerhardt** [comment]
@bartonjs - any thoughts?

**Comment by bartonjs** [comment]
My psychic debugger is getting a weak ping off of libcurl+openssl11 (since I see libssl.so.1.1 installed)

@cobrafast: What curl/libcurl do you have installed?

```sh
$ curl --version
curl 7.47.0 (x86_64-pc-linux-gnu) libcurl/7.47.0 GnuTLS/3.4.10 zlib/1.2.8 libidn/1.32 librtmp/2.3
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP UnixSockets
```

In the case of the machine I happen to be logged in to right now, the TLS provider is `GnuTLS/3.4.10`.  I'm surely hoping yours doesn't say `OpenSSL/1.1(.something)`.  If so, you'll need to get libcurl+openssl1.0(.x) or libcurl+anyotherprovider.

**Comment by cobrafast** [comment]
@bartonjs My curl is newer, but it says `OpenSSL/1.0.2k`.

```
curl 7.52.1 (x86_64-pc-linux-gnu) libcurl/7.52.1 OpenSSL/1.0.2k zlib/1.2.8 libidn2/0.16 libpsl/0.17.0 (+libidn2/0.16) libssh2/1.8.0 nghttp2/1.22.0 librtmp/2.3
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets HTTPS-proxy PSL
```

**Comment by bartonjs** [comment]
Okay, libcurl is the problem.  Or, at least the packages are.  But it's a 1.0 vs 1.0 problem.

At line 224:  `find library=libSystem.Security.Cryptography.Native.OpenSsl.so [0]; searching`

Followed (at 232) by:
```
     10571: find library=libcrypto.so.1.0.0 [0]; searching
     10571:  search cache=/etc/ld.so.cache
     10571:   trying file=/usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0
```

Okay, we've loaded libssl.so.1.0.0.

Then, down at 544: `find library=libSystem.Net.Http.Native.so [0]; searching`

552:
```
     10571: find library=libcurl.so.4 [0]; searching
     10571:  search cache=/etc/ld.so.cache
     10571:   trying file=/usr/lib/x86_64-linux-gnu/libcurl.so.4
```

And.... 580:
```
     10571: find library=libcrypto.so.1.0.2 [0]; searching
     10571:  search cache=/etc/ld.so.cache
     10571:   trying file=/usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.2
```

So we're calling X509_verify in libcrypto.so.1.0.0, but giving it a pointer created by libcrypto.so.1.0.2.  And something between the two different compilations has clearly changed the offset of some value (or interpretation thereof).

I don't know why Debian has packaged them side by side like that.  Your system devel says that "1.0.0" is the SONAME to bind (/usr/lib/x86_64-linux-gnu/libssl.so -> libssl.so.1.0.0).  If there isn't a libcurl which is bound to `libssl.so.1.0.0` then you'll either need to

* Change to libcurl+gnutls (or libcurl+nss, etc)
* Change your development package to be libssl-dev 1.0.2 and rebuild the shims locally.

**Comment by sekjun9878** [comment]
Just to confirm, Debian Stretch does not ship with libssl1.0.0. It was an artifact package required by some userspace apps like skype (libssl1.0.0:i386) and rstudio and ruby2.1 (libssl1.0.0:amd64) in my case. I've removed both of those packages and now I'm getting this error:

```
Unhandled Exception: System.TypeInitializationException: The type initializer for 'Crypto' threw an exception. ---> System.TypeInitializationException: The type initializer for 'CryptoInitializer' threw an exception. ---> System.DllNotFoundException: Unable to load DLL 'System.Security.Cryptography.Native.OpenSsl': The specified module could not be found.
 (Exception from HRESULT: 0x8007007E)
   at Interop.CryptoInitializer.EnsureOpenSslInitialized()
   at Interop.CryptoInitializer..cctor()
   --- End of inner exception stack trace ---
   at Interop.Crypto..cctor()
   --- End of inner exception stack trace ---
   at Interop.Crypto.GetRandomBytes(Byte* buf, Int32 num)
   at System.IO.Path.GetCryptoRandomBytesOpenSsl(Byte* bytes, Int32 byteCount)
   at System.IO.Path.GetCryptoRandomBytes(Byte* bytes, Int32 byteCount)
   at System.IO.Path.GetRandomFileName()
   at Microsoft.DotNet.InternalAbstractions.TemporaryDirectory..ctor()
   at Microsoft.Extensions.EnvironmentAbstractions.DirectoryWrapper.CreateTemporaryDirectory()
   at Microsoft.DotNet.Configurer.NuGetPackagesArchiver..ctor()
   at Microsoft.DotNet.Cli.Program.ConfigureDotNetForFirstTimeUse(INuGetCacheSentinel nugetCacheSentinel)
   at Microsoft.DotNet.Cli.Program.ProcessArgs(String[] args, ITelemetry telemetryClient)
   at Microsoft.DotNet.Cli.Program.Main(String[] args)
Aborted
```

Is dotnet core v1 supposed to support libssl1.0.2 or do I need libssl1.0.0?

**Comment by bartonjs** [comment]
@sekjun9878 If you build locally, any 1.0.x build is fine.  The prebuilts, which were built on/for Jessie, require libssl.1.0.0.  .NET Core 2.0 has added more flexibility, and will work prebuilt on both Jessie and Stretch.

**Comment by omajid** [comment]
It looks like this will be supported in the future: https://github.com/dotnet/corefx/commit/4d7ee843bf1496c0d9c01671befdeb65548aef1f

But the 1.0.x branch links against `libcrypto.so.10` and `libssl.so.10` directly. You can see this via `ldd path/to/dotnet/shared/Microsoft.NETCore.App/1.0.*/System.Security.Cryptography.Native.so`.

**Comment by cobrafast** [comment]
I just tried the .NET Core 2.0 Preview 1 package as per https://www.microsoft.com/net/core/preview#linuxdebian and the problem seems to persist.

`fish: “dotnet restore” terminated by signal SIGSEGV (Address boundary error)`

```
.NET Command Line Tools (2.0.0-preview1-005977)

Product Information:
 Version:            2.0.0-preview1-005977
 Commit SHA-1 hash:  414cab8a0b

Runtime Environment:
 OS Name:     debian
 OS Version:  8
 OS Platform: Linux
 RID:         debian.8-x64
 Base Path:   /opt/dotnet/sdk/2.0.0-preview1-005977/

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.0-preview1-002111-00
  Build    : 1ff021936263d492539399688f46fd3827169983

```

**Comment by ellismg** [comment]
@cobrafast just to be sure, this is on Jessie, right?

**Comment by cobrafast** [comment]
@ellismg it's the same system I've been talking about previously (i.e. https://github.com/dotnet/cli/issues/3681#issuecomment-298802996)

**Comment by bartonjs** [comment]
@cobrafast Support for Jessie was fixed with https://github.com/dotnet/corefx/commit/4d7ee843bf1496c0d9c01671befdeb65548aef1f, but that happened after the preview1 build.  You'll need to try nightlies, or wait for the next release.

**Comment by cobrafast** [comment]
Issue seems to persist with the build from the dotnet debian package repository.

```
deb [arch=amd64] https://packages.microsoft.com/repos/microsoft-debian-stretch-prod stretch main
```

```
fish: “dotnet build” terminated by signal SIGSEGV (Address boundary error)
```

```
#0  0x00007fffe0abfd6d in ?? () from /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0
#1  0x00007fffe0abac8b in X509_verify_cert () from /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0
#2  0x00007fff7d237ff6 in ?? ()
#3  0x00007fff54f1edb0 in ?? ()
#4  0x6c00fa377b77ae7b in ?? ()
#5  0x00000000d0408626 in ?? ()
#6  0x00007ffff67b4488 in ?? () from /usr/share/dotnet/shared/Microsoft.NETCore.App/2.0.0/libcoreclr.so
#7  0x00007fff54f1f688 in ?? ()
#8  0x00007fff7e652100 in ?? ()
#9  0x00007fff7e652100 in ?? ()
#10 0x00007fff54f1edb0 in ?? ()
#11 0x00007fff7d237ff6 in ?? ()
#12 0x00007fff54f1ee40 in ?? ()
#13 0x00007fff7e652100 in ?? ()
#14 0x00007fff5823d308 in ?? ()
#15 0x0000000000000001 in ?? ()
#16 0x00007fff582306a8 in ?? ()
#17 0x00007fff5823d378 in ?? ()
#18 0x00007fff5823d308 in ?? ()
#19 0x00007fff5823d308 in ?? ()
#20 0x00007fff301330c0 in ?? ()
#21 0x00007fff54f1eeb0 in ?? ()
#22 0x00007fff7e43ce57 in ?? ()
#23 0x00007fff54f1ee50 in ?? ()
#24 0x00007fff5823d308 in ?? ()
#25 0x00007fff582306a8 in ?? ()
#26 0x0000000000000000 in ?? ()
```

```
.NET Command Line Tools (2.0.0)

Product Information:
 Version:            2.0.0
 Commit SHA-1 hash:  cdcd1928c9

Runtime Environment:
 OS Name:     debian
 OS Version:  8
 OS Platform: Linux
 RID:         debian.8-x64
 Base Path:   /usr/share/dotnet/sdk/2.0.0/

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.0
  Build    : e8b8861ac7faf042c87a5c2f9f2d04c98b69f28d
```

```
ii  dotnet-host                                                 2.0.0-1                                         amd64
ii  dotnet-hostfxr-2.0.0                                        2.0.0-1                                         amd64
ii  dotnet-runtime-2.0.0                                        2.0.0-1                                         amd64
ii  dotnet-sdk-2.0.0                                            2.0.0-1                                         amd64
ii  libcrypt-openssl-bignum-perl                                0.04-4+b2                                       amd64
ii  libcrypt-openssl-rsa-perl                                   0.28-2+b1                                       amd64
ii  libcurl4-openssl-dev:amd64                                  7.55.1-1                                        amd64
ii  libgnutls-openssl27:amd64                                   3.5.13-2                                        amd64
ii  openssl                                                     1.0.1t-1+deb8u6                                 amd64
```

**Comment by cobrafast** [comment]
Installing packages from the jessie repo doesn't help either.

Perhaps we can get this issue reopened?

**Comment by janvorli** [comment]
@cobrafast can you please run `curl --version` and post the result here? I am interested in seeing the libraries that it shows.

**Comment by cobrafast** [comment]
@janvorli sure:

```
curl 7.55.1 (x86_64-pc-linux-gnu) libcurl/7.55.1 OpenSSL/1.0.2l zlib/1.2.8 libpsl/0.17.0 (+libidn2/0.16) libssh2/1.8.0 nghttp2/1.25.0 librtmp/2.3
Release-Date: 2017-08-14
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets HTTPS-proxy PSL 
```

**Comment by janvorli** [comment]
@cobrafast the curl was built against OpenSSL 1.0.2l and according to your list you seem to have OpenSSL 1.0.1t installed. Do you happen to have the 1.0.2l package installed too? I am not sure about Debian 8, but in Debian 9, the libssl 1.0.2 has a different soname than libssl 1.0.1, those two can be installed side by side and if they end up in the same process (e.g. if one was pulled in by libcurl and the other by .NET native libraries), a threading issues can happen leading to crashes that may be what you are getting.
To see if that's the case, when you run it under gdb and it crashes, could you please run the "info proc map" debugger command and share the output e.g. via gist?

**Comment by cobrafast** [comment]
@janvorli hmm, there are several `libssl` packages installed.

```
ii  libssl-dev:amd64                                            1.0.1t-1+deb8u6                                 amd64
ii  libssl-doc                                                  1.0.1t-1+deb8u6                                 all
ii  libssl0.9.8                                                 0.9.8o-4squeeze14                               amd64
ii  libssl1.0.0:amd64                                           1.0.1t-1+deb8u6                                 amd64
ii  libssl1.0.0:i386                                            1.0.1t-1+deb8u6                                 i386
ii  libssl1.0.2:amd64                                           1.0.2l-2                                        amd64
ii  libssl1.0.2:i386                                            1.0.2l-2                                        i386
ii  libssl1.1:amd64                                             1.1.0f-5                                        amd64
```

Of those I was able to remove `libssl0.9.8` but not the other three (i.e. 1.0.0, 1.0.2, 1.1) since other packages depend on them.

The `info proc map` is here: https://pastebin.com/wpT57znN

**Comment by cobrafast** [comment]
Seems like downgrading `curl` let me remove `libssl1.0.2` and fixed the problem. I hope this doesn't break any other programs though.

**Comment by janvorli** [comment]
The proc map shows I was right - both the libssl.so.1.0.0 and libssl.so.1.0.2 were loaded at the same time. It seems that I'll need to add an env var that would allow people to override the libssl version that we use to allow fixing these corner cases.

**Comment by v6ak** [comment]
* Is there any way to ignore one of the versions, or we have to wait until the next release?
* IIUC, DotNet Core on Debian Jessie is limited to an unsupported version of OpenSSL. (Moreover, this version will not be present in clean Jessie installations.) So, this version can contain vulnerabilities that will never be fixed in Jessie. Are there any plans to upgrade DotNet Core to use the newer OpenSSL version?

In the meantime, I've tried to symlink /usr/lib/x86_64-linux-gnu/libssl.so.1.0.0 to /usr/lib/x86_64-linux-gnu/libssl.so.1.0.2 . **Do it only at your own risk, this might introduce more issues than it fixes.** So far, it looks good, though.

**Comment by janvorli** [comment]
@v6ak on Debian Jessie, the 1.0.1 is the supported version. See https://packages.debian.org/search?suite=jessie&searchon=names&keywords=openssl. The CURL that @cobrafast was using is not the version that's the standard version for Debian Jessie. His version was 7.55 while Debian Jessie has version 7.38. See
https://packages.debian.org/search?suite=jessie&section=all&arch=any&searchon=names&keywords=curl

The Debian Stretch is the one that has moved to 1.0.2.
